lstests/doc/README.md

```diff
@@ -0,0 +1,688 @@
+
+Note that, `buffered=True` only provides a possibility to buffer. Buffering does not happen automatically. A line of information will only be buffered when it is explicitly told  to do so.
+
+For example, given two threads, both scheduled to execute
+
+```python
+# clus is a Cluster object
+clus.log("{}: first half.".format(thread_name), level="debug", buffer=True)
+clus.log("{}: second half.".format(thread_name), level="debug", buffer=True)
+clus.flush()
+```
+
+or equivalently
+
+```python
+# clus is a Cluster object
+clus.log("{}: first half.".format(thread_name), level="debug", buffer=True)
+clus.log("{}: second half.".format(thread_name), level="debug")
+```
+
+setting `buffered=True` will guarantee that the log file will be looking like
+
+```
+thread-1: first half
+thread-1: second half
+thread-2: first half
+thread-2: second half
+```
+
+or
+
+```
+thread-2: first half
+thread-2: second half
+thread-1: first half
+thread-1: second half
+```
+
+but will never be of the following form
+
+```
+thread-2: first half
+thread-1: first half
+thread-1: second half
+thread-2: second half
+```
+
+#### Log Levels
+
+| level   | value |
+| ------- | ----- |
+| `OFF`   | 100   |
+| `FLAG`  | 90    |
+| `FATAL` | 60    |
+| `ERROR` | 50    |
+| `WARN`  | 40    |
+| `INFO`  | 30    |
+| `DEBUG` | 20    |
+| `TRACE` | 10    |
+| `ALL`   | 0     |
+
+The `LogLevel` class accepts either a numeric value or a case-insensitive string for the level. While there are text names given for the level values, all the integers in `[0, 100]` are valid log levels.
+
+```python
+clus.log("Some warning information.", level="warn")
+```
+
+is equivalent to
+
+```python
+clus.log("Some warning information.", level=40)
+```
+
+or
+
+```python
+clus.warn("Some warning information.")
+```
+
+As another example,
+
+```python
+clus.log("Some info more prominent than INFO but not yet a WARN", level=35)
+```
+
+This line will be written to any logging destination with a threshold at or lower than 35. It will be printed out on the screen if the `ScreenWriter` has an `INFO`-level threshold. It will be written to a log file if there is a file writer `FileWriter(threshold=35)`. But it will not appear on the log file that captures only `WARN`-or-higher-level information.
+
+### `Service`
+
+A `Service` object represents the connection with the launcher service running in the background. Once a `Service` object is created, it will try to connect the launcher service automatically.
+
+After configuration, the `Service` will change the working directory, register (and possibly *modify*) the Logger, and then connect to launcher service. *Currently, it is only possible to connect to a local launcher service.*
```

Comment:
Why is it changing the working directory? And how and why is it possibly modifying the Logger? If this isn't necessary, it would simplify the interface for the user if it just didn't mutate any of these things. If it is necessary, then it would be helpful to the reader if the documentation explains this stuff so they aren't confused by why, for example, their Logger configuration has mysteriously changed from what they expected.

Reply:
> Why is it changing the working directory?

I will change the code so that `Service` will not change directory by default. That is, I will make the current working directory the default value for `wdir`. Changing working directory will only happen when users demand so.

> how and why is it possibly modifying the Logger?

This happens only when the user override the logger settings (e.g. change screen logging threshold from `info` to `debug`) from the command line. When the command-line logger settings differ from what they wrote in the script, `Logger` configuration will be changed by the `Service` object. The `Service` has a reference to (instead of another copy of) to the `Logger` object, which keeps all the logging-related settings.

I will make this clearer in the doc.

Reply:
So it seems like the `Service` object owns the `Logger` object and is free to mutate it as it wishes. Is it clear to the user that it gives up ownership of the `Logger` it created to the `Service` it is constructing? I'm concerned about a user creating a `Logger` with some settings, passing it into the constructor of a `Service` (which modifies it according to the command-line overrides), and then the user makes the mistake of thinking the `Logger` they constructed is immutable and can be re-used to construct another `Service` object without realizing that the `Logger` has been mutated.

At the very least this ownership semantics should be clearly documented. But I think it would be better to just make a copy of the `Logger` when constructing the `Service` object to avoid this complexity.

Reply:
> So it seems like the Service object owns the Logger object and is free to mutate it as it wishes.

Yes, it does. The expected use pattern for the launcher-service-based testing framework is:

**One Logger ----> One Service --> Multiple Clusters**

Although we didn't impose hard technical restrictions (like using a singleton pattern) to limit users to only one `Service`, we does expect so. Allowing for construction of multiple `Service` objects brings no added value but will cause unnecessary troubles.

For example, say there are two `Service` objects in `script.py`, and the user runs it with a command-line argument `./script.py --option=a`. It will lead to a confusion: what object(s) would the command-line argument `--option=a` have effect on? On both `Service` object, or just the first one? What about the `Logger` object? Has it been changed?

It is also because of this concern we pass the `Logger` object by reference, not by copy, to the `Service` object, so that the command-line arguments will have a *consistent* effect. That is, if you pass `--monochrome` from the command line, the logger will become a monochrome logger, and the service will log without colors.